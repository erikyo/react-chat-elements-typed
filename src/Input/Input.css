.rce-container-input {
    display: flex;
    gap: 1rem;
    min-width: 100%;
    box-sizing: border-box;
    flex-direction: row;
    background: #f0f2f5;
    padding: 5px 24px;
    align-items: center;
    justify-content: space-between;
}

.rce-input {
    flex: 1;
    height: 42px;
    padding: 5px 12px;
    border: none;
    border-radius: 8px;
    font-size: 15px;
    box-sizing: border-box;
    outline: none;
}

.rce-input-textarea {
    padding: 10px 12px;
    resize: none;
}

.rce-input-buttons {
    display: flex;
    flex-direction: row;
    margin: 5px;
}

.rce-input-buttons > * {
    display: flex;
    flex-direction: row;
}

.rce-input-buttons .rce-button:nth-child(even) {
    margin-left: 5px;
    margin-right: 5px;
}

.rce-input-buttons .rce-button:last-child {
    margin-right: 0;
}

.rce-left-padding {
    padding-left: 10px;
    padding-right: 0 !important;
}

.rce-right-padding {
    padding-right: 10px;
    padding-left: 0 !important;
}

.rce-input::placeholder {
    color: #afafaf;
}
