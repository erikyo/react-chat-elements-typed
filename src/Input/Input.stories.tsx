import type { Meta } from "@storybook/react";
import Input from "./Input";
import type { IInputProps } from "../types";
import { IconEmoji } from "../SvgIcon/IconEmoji";
import { IconPlus } from "../SvgIcon/IconPlus";
import { IconSend } from "../SvgIcon/IconSend";

const meta = {
	component: Input,
} satisfies Meta<typeof Input>;

export default meta;

export const Default = {
	args: {
		multiline: false,
		autoHeight: true,
		leftButtons: (
			<div style={{ display: "flex", gap: 10, justifyContent: "center" }}>
				<IconEmoji style={{ color: "#566570" }} />
				<IconPlus style={{ color: "#566570" }} />
			</div>
		),
		rightButtons: (
			<div style={{ display: "flex", gap: 10, justifyContent: "center" }}>
				<IconSend style={{ color: "#566570" }} />
			</div>
		),
	} as IInputProps,
};
