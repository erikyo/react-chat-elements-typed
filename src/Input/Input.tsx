import type React from "react";
import type { FC, LegacyRef, SyntheticEvent } from "react";
import { useEffect, useState } from "react";
import "./Input.css";
import classNames from "classnames";
import type { IInputProps } from "../types";
import Button from "../Button/Button";
import { MdClear } from "react-icons/md";

const Input: FC<IInputProps> = ({
	type = "text",
	multiline = false,
	minHeight = 25,
	maxHeight = 200,
	autoHeight = true,
	autofocus = false,
	clearButton = true,
	placeholder = "Type here...",
	value: propValue = "",
	style = {},
	...props
}) => {
	const [value, setValue] = useState<string>(
		propValue || props.defaultValue || "",
	);

	useEffect(() => {
		if (autofocus) props.reference?.current?.focus();
	}, [autofocus, props.reference?.current?.focus]);

	/** Update state if propValue changes */
	useEffect(() => {
		setValue(propValue);
	}, [propValue]);

	const onSubmitEvent = (ev: SyntheticEvent<HTMLFormElement, Event>) => {
		ev.preventDefault();
		setValue("");
		if (props.onSubmit instanceof Function) props.onSubmit(ev);
	};

	const onChangeEvent = (
		ev: SyntheticEvent<HTMLInputElement | HTMLTextAreaElement, Event>,
	) => {
		const el = ev.target as HTMLInputElement;
		let newValue = el.value;
		setValue(newValue);

		if (multiline) {
			if (autoHeight) {
				if (el.style.height !== `${minHeight}px`) {
					el.style.height = `${minHeight}px`;
				}

				let height = "";
				if (el.scrollHeight <= maxHeight) height = `${el.scrollHeight}px`;
				else height = `${maxHeight}px`;

				if (el.style.height !== height) {
					el.style.height = height;
				}
			}
		}

		if (props.maxlength && (el.value || "").length > props.maxlength) {
			if (props.onMaxLengthExceed instanceof Function)
				props.onMaxLengthExceed();

			if (props.reference?.current) {
				newValue = newValue.substring(0, props.maxlength);
				(props.reference.current as HTMLInputElement).value = newValue;
				setValue(newValue);
			}
			return;
		}
	};

	const clear = () => {
		if (props.reference?.current) {
			(props.reference.current as HTMLInputElement).value = "";
		}

		setValue("");
	};

	return (
		<form
			onSubmit={onSubmitEvent}
			className={classNames("rce-container-input", props.className)}
			style={style}
		>
			{props.leftButtons && (
				<div className="rce-input-buttons">{props.leftButtons}</div>
			)}
			<div className="rce-input-content w-full relative">
				{!multiline ? (
					<input
						name={"chat-input"}
						ref={props.reference as LegacyRef<HTMLInputElement>}
						type={type}
						className={classNames("rce-input w-full")}
						placeholder={placeholder}
						value={value}
						style={props.inputStyle}
						onChange={(e) => onChangeEvent(e)}
						onCopy={props.onCopy}
						onCut={props.onCut}
						onPaste={props.onPaste}
						onBlur={props.onBlur}
						onFocus={props.onFocus}
						onSelect={props.onSelect}
						onReset={props.onReset}
						onKeyDown={props.onKeyDown}
						onKeyUp={props.onKeyUp}
					/>
				) : (
					<textarea
						ref={props.reference as LegacyRef<HTMLTextAreaElement>}
						className={classNames("rce-input", "rce-input-textarea")}
						placeholder={placeholder}
						style={props.inputStyle}
						onChange={(e) => onChangeEvent(e)}
						onCopy={props.onCopy}
						onCut={props.onCut}
						onPaste={props.onPaste}
						onBlur={props.onBlur}
						onFocus={props.onFocus}
						onSelect={props.onSelect}
						onReset={props.onReset}
						onKeyDown={props.onKeyDown}
						onKeyUp={props.onKeyUp}
					>
						{value}
					</textarea>
				)}
				{clearButton && value && (
					<Button
						className="rce-input-clear"
						onClick={clear}
						backgroundColor={"transparent"}
						style={{ position: "absolute", zIndex: 1, right: 0, opacity: 0.3 }}
						icon={{
							component: <MdClear color={"var(--rce-color-gray)"} size={20} />,
						}}
					/>
				)}
			</div>
			{props.rightButtons && (
				<div className="rce-input-buttons">{props.rightButtons}</div>
			)}
		</form>
	);
};

export default Input;
