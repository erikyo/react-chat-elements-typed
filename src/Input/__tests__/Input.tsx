import { describe, expect, it } from "vitest";
import React from "react";
import Input from "../Input";
import { render } from "@testing-library/react";

describe("Avatar component", () => {
	it("should render without issues", () => {
		const component = render(<Input />);
		expect(component.container).toBeTruthy();
		expect(component.container).toMatchSnapshot();
	});
});
