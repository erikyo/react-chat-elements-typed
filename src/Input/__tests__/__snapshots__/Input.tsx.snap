// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar component > should render without issues 1`] = `
<div>
  <form
    class="rce-container-input"
  >
    <div
      class="rce-input-content w-full relative"
    >
      <input
        class="rce-input w-full"
        name="chat-input"
        placeholder="Type here..."
        type="text"
        value=""
      />
    </div>
  </form>
</div>
`;
